### DISCUSSION TITLE
Add more built-in components to Starlight

### DISCUSSION THREAD

**Discussion by delucis** [withastro/starlight#1491] [Feature Requests]
We have several components we built specifically for Astro’s Docs or ideas for components that we could add that would likely be generally helpful to Starlight users as they meet common documentation needs. This discussion is a place to share these ideas and give feedback.

### How to use this discussion

- Upvote components that would be useful to you 🗳️ 
- Add additional thoughts as replies to the comment for the relevant component 💬 
- Add an idea for a different component as a separate comment 💡

**Comment by delucis** [comment]
## `<FileTree>`

The Astro and Starlight docs both use a file tree component for rendering directory structures. For example, in Starlight’s [“Sidebar”](https://starlight.astro.build/guides/sidebar/) guide:

<img width="741" alt="Directory structure featuring icons for file types and collapsible directories" src="https://github.com/withastro/starlight/assets/357379/ecf09f56-14aa-4eeb-b441-e19f3a003a27">

We already have a PR for adding this component to Starlight here: https://github.com/withastro/starlight/pull/1308

**Reply by bdenham** [reply]
I implemented a simpler, more compact version for my own usage, so I’m keen to see a native version implemented. That will be one less custom component I need to worry about!

**Reply by ematipico** [reply]
Very much needed!

**Reply by HiDeoo** [reply]
Update: the [`<FileTree>` component](https://starlight.astro.build/guides/components/#file-tree) is available in Starlight v0.21 :tada:

**Comment by delucis** [comment]
## `<Aside>`

Starlight supports notes, tips, and other asides via a [custom Markdown syntax](https://starlight.astro.build/guides/authoring-content/#asides):

```md
:::tip
This is a tip
:::
```

In some contexts it could be helpful to have access to a `.astro` component version of this (e.g. to configure Markdoc or when using a custom page as will soon be possible via https://github.com/withastro/starlight/pull/1175)

**Reply by delucis** [reply]
Update on this one: the [`<Aside>` component](https://starlight.astro.build/guides/components/#asides) is available in Starlight v0.19!

**Comment by delucis** [comment]
## `<Badge>`

Starlight already includes the ability to add badges to sidebar items. For example, in the current Starlight docs sidebar:

<img width="284" alt="navigation sidebar section with a “New” badge on the category title" src="https://github.com/withastro/starlight/assets/357379/a6273519-9f53-42c6-914e-cc1d1494a2a1">

Would it be useful for people to be able to use this more freely inside content?

**Reply by HiDeoo** [reply]
Definitely one of the component I have duplicated the most in plugins: [starlight-obsidian](https://github.com/HiDeoo/starlight-obsidian/blob/main/packages/starlight-obsidian/components/Tags.astro) - [starlight-blog](https://github.com/HiDeoo/starlight-blog/blob/main/packages/starlight-blog/components/PostTags.astro)

**Reply by kevinzunigacuellar** [reply]
I would like to champion this one, if possible 🥇

**Reply by delucis** [reply]
Thanks to @kevinzunigacuellar in https://github.com/withastro/starlight/pull/1530 the [`<Badge>` component](https://starlight.astro.build/guides/components/#badges) is available in Starlight v0.24 and higher!

**Comment by delucis** [comment]
## `<Button>` (call to action)

Starlight’s hero component includes call-to-action style buttons:

<img width="422" alt="two call-to-action links, one styled with a solid background colour, the other a more minimal text link" src="https://github.com/withastro/starlight/assets/357379/2e80ddd6-e193-416a-9a7a-c61ed5b9e3e2">

Would it be useful to be able to use these in content?

**Reply by vvs-basanth** [reply]
Would it be possible to introduce a feature that allows us to customize the opacity of buttons when hovered over? This enhancement would enable us to apply specific styling properties, providing greater control over the visual feedback when users interact with buttons.

This addition would be highly beneficial for achieving a more tailored user experience.

**Reply by HiDeoo** [reply]
Update: the [`<LinkButton>` component](https://starlight.astro.build/components/link-buttons/) is available in Starlight [v0.26](https://github.com/withastro/starlight/releases/tag/%40astrojs%2Fstarlight%400.26.0) 🎉

**Comment by delucis** [comment]
## `<MultipleChoice>`

Astro’s [“Build a Blog” tutorial](https://docs.astro.build/en/tutorial/1-setup/1/#test-your-knowledge) features multiple choice quizzes used for interactive tests of a user’s knowledge:

<img width="671" alt="multiple choice quiz with three options and a result reading “success”" src="https://github.com/withastro/starlight/assets/357379/c9ad3088-305e-4408-946f-cd3c3ac2e95a">

The authoring experience is quite nice on these:

```mdx
<MultipleChoice>
  <Option>
    web browser
  </Option>
  <Option>
    Terminal
  </Option>
  <Option isCorrect>
    VS Code
  </Option>
</MultipleChoice>
```

Maybe people are interested in getting access to this as part of Starlight?

**Reply by lschierer** [reply]
more generically, support for Astro's [HML Forms](https://docs.astro.build/en/recipes/build-forms/)

**Comment by delucis** [comment]
## `<Steps>`

[Nextra](https://nextra.site/docs/guide/built-ins/steps) and [Vocs](https://vocs.dev/docs/markdown#steps) both include a `<Steps>` component for rendering a more in-your-face numbered list of steps.

Would that be something people would find helpful in their Starlight sites?

**Reply by sarah11918** [reply]
I am in favour of this one! We are constantly finding that we need to fiddle with indentation to have code and instructions inside numbered steps. We are constantly getting it wrong, then having other Docs PRs to clean it up.

We'd still have to do that inside a component, BUT I feel like it would be "a thing" and easier to document and make sure people get right all the time.

**Reply by bdenham** [reply]
This is one that I implemented on my own as an Astro component. Would be nice to see it natively implemented. Again, one less custom component required.

**Reply by HiDeoo** [reply]
Update: the [`<Steps>` component](https://starlight.astro.build/guides/components/#steps) is available in Starlight v0.21 :rocket:

**Reply by larkiny** [reply]
There is a minor bug with the `Steps` component. If you include an unordered list inside of the Steps component, it will count those list items as if they were part of the parent ordered list. It should ignore any `<li>` elements that are not direct children of the ordered list. There are also some oddities that trigger a runtime error. For example, sometimes when I have an `Image` component inside the `Steps` component, a runtime error is thrown complaining that there can only be one child `<ol>` element, but if I bring the image component inline with text, the error does not get triggered.

**Reply by lorenzolewis** [reply]
@larkiny which version of Starlight are you using? This was just recently fixed and released in 0.24.4 https://github.com/withastro/starlight/issues/2041

Edit: fixed version number typo

**Reply by HiDeoo** [reply]
As mentioned by Lorenzo, this should now be fixed in [version `0.24.4`](https://github.com/withastro/starlight/releases/tag/%40astrojs%2Fstarlight%400.24.4) but if you are still facing the problem, feel free to open a new issue with a reproducible example including the browser version and operating system you are using.

Regarding the second point, the error message has been improved in Starlight 0.22.3 and should now tell you which elements have been detected as invalid, e.g.:

> The `<Steps>` component expects its content to be a single ordered list (`<ol>`) but found multiple child elements: `<ol>`, `<img>`.

If you're seeing the above error with an `<img>` element, the issue sounds like an indentation problem, e.g. the following code:

```mdx
<Steps>

1. First step

2. Second step with an image

<Image src={Logo} alt="Logo" />

</Steps>
```

The correct syntax would be:

```mdx
<Steps>

1. First step

2. Second step with an image

   <Image src={Logo} alt="Logo" />

</Steps>
```

Notice the indentation of the `<Image>` component which should be inside the second step and follow the Markdown [rule](https://www.markdownguide.org/basic-syntax/#adding-elements-in-lists) for adding elements in lists.

**Reply by larkiny** [reply]
It was indeed an indentation problem, so that one is on me. Sorry! I fixed the nested list item problem with some CSS overrides. I am indeed on Starlight 0.23.2, so I need to upgrade. Thanks for the quick help!

**Comment by santiemanuel** [comment]
## `<ImageComponent>`

An image component with the ability to expand on click. Some images we use in Starlight are dense in information and those are only good for desktop users. Having the ability to zoom in an image would help mobile users to read the info.

**Reply by bdenham** [reply]
Yes, please! This is a big one for me. I try to include at least one diagram for every topic and it can be really hard to design it in a way that the callout text and graphics fit within the ~720px width of the content area for a page—while remaining readable!

In addition, I would also like to see an option for entering a caption that would appear consistently below the image. 

I’d also like to see the same thing for a `<VideoComponent />`.

**Reply by libreom** [reply]
I need this for my doc site also thanks

**Reply by HiDeoo** [reply]
Not exactly a component as discussed here, but as a potential alternative, I made a Starlight [plugin](https://github.com/HiDeoo/starlight-image-zoom) adding zoom capabilities to documentation images in case this might be useful to someone.

**Reply by bdenham** [reply]
@HiDeoo, I just went through your docs for the plugin. It’s perfect! The implementation and decisions you made for its design and default feature set are all spot on for my needs. Extremely well done. You made my week! Bravo!

**Comment by prateek-tailor** [comment]
# `<Code>`

<br>

An `expressive-code` component that can be used to wrap additional functionality above/below the code block within markdown while maintaining a unified experience with expressive-code's look and feel. 

## What could this enable? 
* atomically build custom UI components using `<Code>`
* Easier implementation of two panel docs (left side content, right side code blocks)
* Ability to build language selector above code blocks
* For HTML code blocks, ability to render the html above the code block 
* For GQL code blocks, ability to add a button below a code block to "run in playground"


## Examples
* flex-box with code blocks: https://graphql.org/learn/queries/#using-variables-inside-fragments
* two panel design: https://shopify.dev/docs/api/admin-graphql/2024-01/mutations/marketLocalizationsRegister
* lang selector: https://stripe.com/docs/payments/accept-a-payment?platform=web&ui=checkout&lang=ruby#set-up-stripe
* HTML rendered above code blocks: https://mui.com/material-ui/react-button-group/#basic-button-group

**Reply by delucis** [reply]
We have great news for you! A [`<Code>` component](https://starlight.astro.build/guides/components/#code) is available as of v0.17 of Starlight

**Comment by Darkflame72** [comment]
# `<ListCard>` and `<SplitCard>`

Add the `ListCard` and `SplitCard` Cards that are used on the [Astro Docs](https://docs.astro.build/en/getting-started/).

These provide more options with cards and styles for how they are used. List cards are often used to link to various parts of the docs and the Split card can be used for installation details or other important information.

<img width="1222" alt="Screenshot 2024-03-29 at 11 08 26 PM" src="https://github.com/withastro/starlight/assets/31436575/47ad9212-703b-45fe-8ccc-682271a76f5e">

**Reply by at-the-vr** [reply]
I mean [Cards](https://starlight.astro.build/guides/components/#cards) and [Link Cards](https://starlight.astro.build/guides/components/#link-cards) pretty much do the thing that your looking for, just the markup and css is left for user to manage

**Reply by Darkflame72** [reply]
They are similar but I can very much see use cases for these specific cards especially with a collection of links rather than a big block link.

**Comment by Darkflame72** [comment]
# `<Math>`

Add a block to render [MathJax](https://www.mathjax.org) this would be similar to the opt in functionality in [VitePress](https://vitepress.dev/guide/markdown#math-equations). This is especially useful for any documentation around Machine Learning/Artifical Intelligence, Computer Graphics, Computer Vision and more where equations should be rendered near the code

**Reply by HiDeoo** [reply]
Just like VitePress uses `markdown-it-mathjax3` to enable these features, Starlight is built on top of Astro which supports adding third-party Markdown plugins for Markdown and MDX.

In this specific case, this would be [`remark-math`](https://github.com/remarkjs/remark-math/tree/main/packages/remark-math) and [`rehype-mathjax`](https://github.com/remarkjs/remark-math/tree/main/packages/rehype-mathjax).

You can find more informations with a small example and links to more example in this [post](https://github.com/withastro/starlight/issues/721#issuecomment-1722290339). The entire conversation (and the last message) also contains various information about styling such expressions which you might find useful.

**Reply by Darkflame72** [reply]
It might be worth documenting this on the starlight docs as it may not be immediatly obvious (at least wasn't too me).

**Reply by torn4dom4n** [reply]
I think we need a plugin more than a built-in component. BTW, I'm using KaTeX to render LaTeX in starlight. You can find it [here](https://area44.github.io/playastro/guides/latex/).

**Reply by stereobooster** [reply]
Here is how to configure math support in Starlight https://astro-digital-garden.stereobooster.com/recipes/math-support-in-markdown/

**Reply by delucis** [reply]
Here's another example of a Starlight project that can render MathJax syntax: https://starlight-examples.netlify.app/examples/mathjax/

**Comment by stereobooster** [comment]
## `<kbd>`

Not exactly an Astro component, but rather CSS for HTML tag (kbd = keyboard)

Example:

```md
<kbd>Cmd</kbd> + <kbd>F</kbd>
```

<kbd>Cmd</kbd> + <kbd>F</kbd>

**Reply by at-the-vr** [reply]
Decided to spin up a quick starlight site for this one <https://htkbd.vercel.app/> for anyone interested

**Reply by trueberryless** [reply]
I think this can be marked as complete, HiDeoo recently developed a plugin for that: https://github.com/HiDeoo/starlight-kbd

**Comment by spokospace** [comment]
would be great if there was full support for UnoCSS, which is much more convenient than Tailwind itself.
https://unocss.dev/integrations/astro

**Comment by WhyFenceCode** [comment]
In addition to the test/quiz form that was mentioned above, I think there could be a use for a steps component, but that it only shows the first step, and so when you complete a quiz in it, then it shows the second one.

**Comment by renphysics** [comment]
Feature request: Top navigation bar with dropdown menus (on the left or right of the search box)

**Comment by Henry-Hiles** [comment]
Feature request:  LinkButton styling on regular buttons!

**Comment by redxXxkiller** [comment]
# `<ChangeLog>`
Changelog component similar to `<Steps>` but with better design

Example from [docmd](https://docs.docmd.io/content/containers/changelogs/)


<img width="927" height="586" alt="Screenshot 2026-03-08 215510" src="https://github.com/user-attachments/assets/e1bfb5b7-c726-415f-ae4c-a18c8ccf066b" />
